### PR TITLE
Datasource URL extension.

### PR DESCRIPTION
**Pull request by emtwo**

### CODE REVIEW

**Comment by emtwo** [comment]
Note: for the python portion of the extension, a new method, `PostgreSQL.add_configuration_property` needs to be added to the `redash` codebase that is used as an extension hook.

**Comment by emtwo** [comment]
Note: This still needs other data source URLs to be added.

**Comment by emtwo** [comment]
Also note: this is a port of https://github.com/mozilla/redash/commit/75228b214fb36746154c00aebee7d2918d5360bb

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ dist
 *.pyc
 .DS_Store
 .pytest_cache/
+.DS_Store
```
Comment (jezdez) [inline .gitignore]: Duplicate from line 12.

File: setup.py [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
+# Note: package_data does not allow files to be listed recursively,
+# so this helper function is used.
+def get_web_files():
+    matches = []
+    PREFIX = './src/redash_stmo/'
+    for root, dirnames, filenames in os.walk(PREFIX):
+        for filename in filenames:
+            if filename.endswith(".js") or filename.endswith(".html"):
+                matches.append(os.path.join(root, filename).split(PREFIX, 1)[1])
+    return matches
```
Comment (jezdez) [inline setup.py]: Instead of manually building this here, we can use a different feature of setuptools: https://python-packaging.readthedocs.io/en/latest/non-code-files.html

Basically passing `include_package_data=True` to the `setup()` function is enough and then specifying the actual non-Python files in the [`MANIFEST.in` file](https://docs.python.org/2/distutils/sourcedist.html#the-manifest-in-template) (read: file manifest template). It supports globs and recursive inclusion, for our case:

```
include AUTHORS.rst CHANGELOG.rst README.rst
recursive-include src/redash_stmo *.html *.js
```

File: src/redash_stmo/datasource_link.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+from redash.models import DataSource
```
Comment (jezdez) [inline src/redash_stmo/datasource_link.py:1]: Please stay consistent with the 4 character indentation.

File: src/redash_stmo/datasource_link.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	"presto": "https://prestodb.io/docs/current/",
+	"python": "http://redash.readthedocs.io/en/latest/datasources.html#python",
+	"insecure_script": "http://redash.readthedocs.io/en/latest/datasources.html#python",
+	"sqlite": "http://sqlite.org/lang.html",
+	"treasuredata": "https://docs.treasuredata.com/categories/hive",
+	"url": "http://redash.readthedocs.io/en/latest/datasources.html#url",
+	"vertica": (
+		"https://my.vertica.com/docs/8.0.x/HTML/index.htm#Authoring/"
+        "ConceptsGuide/Other/SQLOverview.htm%3FTocPath%3DSQL"
+        "%2520Reference%2520Manual%7C_____1")
```
Comment (jezdez) [inline src/redash_stmo/datasource_link.py]: Linebreak after the string please.

File: setup.py [hunk 5]
```diff
@@ -21,7 +36,11 @@
     entry_points={
         'redash.extensions': [
             'dockerflow = redash_stmo.dockerflow:dockerflow',
-            'datasource_health = redash_stmo.health:datasource_health'
+            'datasource_health = redash_stmo.health:datasource_health',
+            'datasource_link = redash_stmo.datasource_link:datasource_link'
+        ],
+        'webpack.bundles': [
+            'datasource_link = redash_stmo.js_extensions:datasource_link',
```
Comment (jezdez) [inline setup.py]: So looking at this I noticed a problem I hadn't seen when I reviewed the extension API code into our fork, that the need to provide config values for `extension_directory` and `frontend_content` isn't actually really there and looks like redundant. Both values can be calculated from the entrypoint object that is returned when we load the `redash.extensions` entrypoints.

- `extension_directory` -- is the filesystem path of the package `redash_stmo.datasource_link`
- `frontend_content` -- has no need to be defined individually per extension and should instead be consistent for every extension

So I believe we should simplify the structure of an extension further:

#### An extension is a Python module or package which may or may not have a `bundle` subdirectory, and is defined in the `redash.extensions` entrypoint.

The code that loads the bundles in [bundle-extensions](https://github.com/mozilla/redash/blob/9424dadcc7c0e13d9f094b18914615bae756dd92/bin/bundle-extensions#L25-L30) should calculate the original source of the extension frontend files from the entry in the `redash.extensions` entrypoint and use the name of the loaded entrypoint (e.g. `datasource_link`) as the subdirectory in `client/app/extensions`.

That basically would remove the need for the `webpack.bundles` entrypoint again and consolidate frontend and backend into one location.

File: src/redash_stmo/datasource_link.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+from redash.models import DataSource
+from redash.query_runner import BaseQueryRunner, query_runners
+from redash.query_runner import get_configuration_schema_for_query_runner_type
```
Comment (jezdez) [inline src/redash_stmo/datasource_link.py]: You should use parentheses to combine those two imports in one import call.

File: src/redash_stmo/datasource_link.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+		if runner_type not in DATASOURCE_URLS:
+			continue
+
+		runner_class.default_doc_url = DATASOURCE_URLS[runner_type]
+		runner_class.add_configuration_property("doc_url", {
+	        "type": "string",
+	        "title": "Documentation URL",
+	        "default": runner_class.default_doc_url})
+
+	DataSource.to_dict = datasource_to_dict
```
Comment (jezdez) [inline src/redash_stmo/datasource_link.py]: So a few things, adding the configuration property to the runner_class makes sense, although the code is missing so I'm not sure, but the other part where it monkey-patches multiple classes to pass a few strings to a template is not a good idea maintenance wise.

I suggest to instead create a new API endpoint `/api/data_sources/<data_source_id>/link` like so (and do the route registration inside the `datasource_link` entrypoint callback):

```python
from redash.handlers.api import api

class DataSourceLinkResource(BaseResource):
    def get(self, data_source_id):
        data_source = get_object_or_404(
            DataSource.get_by_id_and_org,
            data_source_id,
            self.current_org,
        )
        return {
            'type_name': data_source.query_runner.name()
            'doc_url': data_source.options.get('doc_url', None)
        }


api.add_org_resource(DataSourceLinkResource, '/api/data_sources/<data_source_id>/link')
```

Then in the frontend code we simply fetch the data from that endpoint and pass it to the template when the site is loaded. That way we separate the concerns and don't risk overlaps with *other* extensions in the future that also want to add stuff to the query editor page. FWIW I'm not worried about the extra call to the server since we're not really dealing with a ton of traffic.

File: src/redash_stmo/datasource_link/web/datasource-link.html [hunk 8]
```diff
@@ -0,0 +1,2 @@
+<a ng-if="dataSource.options.doc_url != '' && dataSource.options.doc_url" ng-href={{dataSource.options.doc_url}}>{{dataSource.type_name}} documentation</a>
+<span ng-if="dataSource.options.doc_url == '' || !dataSource.options.doc_url">{{ dataSource.type_name }} documentation</span>
```
Comment (jezdez) [inline src/redash_stmo/datasource_link/web/datasource-link.html]: This needs a tiny bit of update given the difference in what the datasource link API endpoint returns.

File: src/redash_stmo/datasource_link/web/index.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import template from './datasource-link.html';
+
+function controller($scope, $route) {
+  const dataSources = $route.current.locals.dataSources;
+  const query = $route.current.locals.query;
+  $scope.dataSource = dataSources.find(ds => ds.id === query.data_source_id);
```
Comment (jezdez) [inline src/redash_stmo/datasource_link/web/index.js]: I understand that we use Angular's resource thing here to get the data from the API, and I'm not sure how to write the actual API call to the new datasource link API endpoint. But since we'll be doing React in the future, might as well not lean too much on Angular IMO.

In fact, don't you want to try writing this in React instead? Allen could help you get up to speed how to do that and the React setup is already in Redash enabled.

**Review by jezdez — COMMENTED** [review]
Some minor nits, but also two bigger things regarding the extension API design that I only now understand and a no to monkey-patches for this use case since there is a more scalable way.

**Comment by jezdez** [comment]
That said, this PR kind of blows me away in what we'll be able to do with this extension API!

Dat me: 1462065207_wxvke

File: src/redash_stmo/datasource_link.py (hunk 7 above)
Comment (emtwo) [reply src/redash_stmo/datasource_link.py]: Good point! I did not consider an API call as an alternative here, but I like it. Thank you!

To clarify, `add_configuration_property()` would be implemented in `BaseQueryRunner` and it would look something like this:

```
class BaseQueryRunner(object):
    ....
    configuration_properties = None
    ....
    @classmethod
    def add_configuration_property(cls, property, value):
        if cls.configuration_properties is None:
            raise NotImplementedError()
        cls.configuration_properties[property] = value
```

And every class that inherits from `BaseQueryRunner` that wants to add a URL to its config needs to have a `configuration_properties` field. Does that sound reasonable?

Comment (jezdez) [reply src/redash_stmo/datasource_link.py]: Yep, that sounds totally reasonable! Thank you :)

File: setup.py (hunk 5 above)
Comment (emtwo) [reply setup.py]: What about when an extension is frontend only with no python portion? Would we then have an empty python file/function for it to be loaded here: https://github.com/mozilla/redash/blob/master/redash/extensions.py?

Alternatively, `extensions.py` could be modified to skip over empty extensions. 

I think this is originally why I had a separate entry, `webpack.bundles` here. Since not all frontend extensions will have a backend portion.

Comment (jezdez) [reply setup.py]: For pure-frontend extensions the process would still be having a package with its name, e.g.

```
src
└── redash_stmo
    ├── __init__.py
    ├── dockerflow.py
    ├── health.py
    ├── whimsycorn
    │   ├── __init__.py
    │   └── bundle
    │   │   ├── index.js
    │   │   ├── style.css
    │   │   └── template.html
    ├── query_runner
    │   ├── __init__.py
    │   └── activedata.py
    └── settings.py
```

That would allow us:

a) use the Python import system to look for the path of the packages

b) provide a location of extra extension metadata later on (e.g. verbose name, description text etc)

Skipping over extensions by checking if the loaded entrypoint is a callable would make sense. IIRC the Entrypoint object has some good APIs to see what it is.

E.g. the syntax for entrypoints only requires a modulename (read: dotted path to a module or package): https://github.com/pypa/setuptools/blob/01fd232d95ad51f784de53c0993a7b91868d269d/pkg_resources/__init__.py#L2361-L2368

You can use the [`Entrypoint.resolve`](https://github.com/pypa/setuptools/blob/01fd232d95ad51f784de53c0993a7b91868d269d/pkg_resources/__init__.py#L2327-L2335) function which will try to import it and raise an `ImportError` if it fails when you skip those that don't actually have an entrypoint provided.

So an entrypoint list in setup.py could look like this (where the whimsycorn extension just happens to not need a callable):

```python
entrypoints={
    'redash.extensions': [
        'dockerflow = redash_stmo.dockerflow:dockerflow',
        'datasource_health = redash_stmo.health:datasource_health',
        'datasource_link = redash_stmo.datasource_link:datasource_link',
        'whimsycorn = redash_stmo.whimsy_corn',
   ], 
}
```

In the bundle-extensions script you'll repeat some of the logic of course since you'll need to check if the bundle directory exists in the extension directory.

File: src/redash_stmo/datasource_link.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            continue
+
+        runner_class.add_configuration_property("doc_url", {
+            "type": "string",
+            "title": "Documentation URL",
+            "default": DATASOURCE_URLS[runner_type]})
+
+    # After api.init_app() is called, api.app should be set by Flask (but it's not) so that
+    # further calls to add_resource() are handled immediately for the given app.
+    api.app = app
```
Comment (jezdez) [inline src/redash_stmo/datasource_link.py:67]: Is this an issue in Redash or something that Flask-Restful does wrong?

Comment (emtwo) [reply src/redash_stmo/datasource_link.py:67]: This is an issue in Flask-Restful. There are a few things going on that led me to adding this line.

tl;dr: Flask-Restful does not play well with the Redash setup of multiple workers each having a different instance of a flask app.

1. Inside `create_app()` in `redash/__init__.py`, we cannot move `handlers.init_app(app)` (which calls `api.init_app(app)`) after `extensions.init_extensions(app)` because the datasource_health extension overwrites the `status_api` endpoint after it’s created. If extensions initialize first then an attempt is made in Redash to add `status_api` and it crashes. So we keep the init order so redash-stmo can override endpoints that already exist.

2. Unlike `api = Api(app)` (https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L92), `api.init_app()` does not set `api.app` when it's called (https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L196). This results in any further calls to `api.add_resource()` after `api.init_app()` to actually have no effect. It results in appending to `api.resources` (https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L384) but no call to `register_view()`

3. Even if we did manage to work around this issue in `1.`, this would not resolve this issue. A Flask app is created in each worker with `create_app()`. If an extension tries to do `api.add_org_resource()`, the same resource will be added to `api.resources` from multiple workers. Now when `api.init_app(app)` is called, there is an exception thrown that there are multiple view functions with the same name.

The crux of the problem is that `api.resources` is shared among multiple apps as a temporary cache until `api.init_app(app)` is called so that resources can be attached to a given app. Since we have a different app for each worker, and each worker is adding the same resource in an extension, this model doesn't quite work.

By setting `api.app = app` we ensure that the resource is added to the correct app and there is no duplication by caching in api.resources.

Comment (jezdez) [reply src/redash_stmo/datasource_link.py:67]: That makes sense to me!

File: src/redash_stmo/datasource_link/bundle/datasource_link.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      type_name: '',
+      doc_url: '',
+    };
+  }
+
+  _loadURLData() {
```
Comment (jezdez) [inline src/redash_stmo/datasource_link/bundle/datasource_link.js]: Any reason to have this prefixed with an underscore?

File: src/redash_stmo/datasource_link/bundle/datasource_link.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    fetch(`${this.props.clientConfig.basePath}api/data_sources/${this.props.datasourceId}/link`)
+      .then((response) => {
+        if (response.status === 200) {
+          return response.json();
+        }
+        return {};
+      })
+      .then((json) => {
+        const { type_name, doc_url } = json.message;
```
Comment (jezdez) [inline src/redash_stmo/datasource_link/bundle/datasource_link.js]: Forgive my noobness with modern JS, wouldn't this code fail in case the fetch call returns an empty object because for some error happening server side?

I mean, `json.message` would not exist and then the variable unpacking would fail. Does this require more defensive code?

**Review by jezdez — APPROVED** [review]
Awesome, nice work @emtwo!

rwc+

File: src/redash_stmo/datasource_link/bundle/datasource_link.js (hunk 12 above)
Comment (emtwo) [reply src/redash_stmo/datasource_link/bundle/datasource_link.js]: Good catch. I will fix this prior to merging, thanks!

File: src/redash_stmo/datasource_link/bundle/datasource_link.js (hunk 11 above)
Comment (emtwo) [reply src/redash_stmo/datasource_link/bundle/datasource_link.js]: This was just to indicate it's not a built-in React function like the rest. I can remove it if it's more confusing than useful. It doesn't necessarily seem to be a common practice, I've just seen it in a couple of places.

Comment (jezdez) [reply src/redash_stmo/datasource_link/bundle/datasource_link.js]: I don't have a strong opinion either way, just noticed and wondered about the best practice in modern JS is.
